scripts/lib/CIME/XML/env_mach_specific.py

```diff
@@ -160,11 +119,156 @@ def make_env_mach_specific_file(self, shell, case):
+        best_match = None
+        best_num_matched = -1
+        default_match = None
+        best_num_matched_default = -1
+        args = []
+        for mpirun_node in mpirun_nodes:
+            xml_attribs = self.attrib(mpirun_node)
+            all_match = True
+            matches = 0
+            is_default = False
+
+            for key, value in attribs.items():
+                expect(key in self._allowed_mpi_attributes, "Unexpected key {} in mpirun attributes".format(key))
+                if key in xml_attribs:
+                    if xml_attribs[key].lower() == "false":
+                        xml_attrib = False
+                    elif xml_attribs[key].lower() == "true":
+                        xml_attrib = True
+                    else:
+                        xml_attrib = xml_attribs[key]
+
+                    if xml_attrib == value:
+                        matches += 1
+                    elif key == "mpilib" and value != "mpi-serial" and xml_attrib == "default":
+                        is_default = True
+                    else:
+                        all_match = False
+                        break
+
+            if all_match:
+                if is_default:
+                    if matches > best_num_matched_default:
+                        default_match = mpirun_node
+                        best_num_matched_default = matches
+                else:
+                    if matches > best_num_matched:
+                        best_match = mpirun_node
+                        best_num_matched = matches
+
+        # if there are no special arguments required for mpi-serial it need not have an entry in config_machines.xml
+        if "mpilib" in attribs and attribs["mpilib"] == "mpi-serial" and best_match is None:
+            return "",[]
+
+        expect(best_match is not None or default_match is not None,
+               "Could not find a matching MPI for attributes: {}".format(attribs))
+
+        the_match = best_match if best_match is not None else default_match
+
+        # Now that we know the best match, compute the arguments
+        if not exe_only:
+            arg_node = self.get_optional_child("arguments", root=the_match)
+            if arg_node is not None:
+                arg_nodes = self.get_children("arg", root=arg_node)
+                for arg_node in arg_nodes:
+                    arg_value = transform_vars(self.text(arg_node),
+                                               case=case,
+                                               subgroup=job,
+                                               default=self.get(arg_node, "default"))
+                    args.append(arg_value)
+
+        exec_node = self.get_child("executable", root=the_match)
+        expect(exec_node is not None,"No executable found")
+        executable = self.text(exec_node)
+
+        return executable, args
+
+    def get_type_info(self, vid):
+        return "char"
 
     # Private API
```

Comment:
This comment did not separate the public from private parts of the class, now it does.